File: Sources/Runtime/Renderer/ExpressionResolver.swift
import SwiftUI
import Core

/// Extracts Lua expression resolution from component rendering.
/// All typed resolve methods share the same `vm` + `props` dependency pair.
struct ExpressionResolver {
    let vm: LuaVM?
    let props: [String: LuaValue]?
    let state: [String: LuaValue]?

    // MARK: - Internal Helpers

    /// Returns a `local props = { ... }` Lua snippet, or "" if no props.
    func propsPrefix() -> String {
        guard let vm, let props else { return "" }
        return vm.localPrefix(key: "props", for: props)
    }

    /// Returns a `local state = { ... }` Lua snippet, or "" if no props.
    func statePrefix() -> String {
        guard let vm, let state else { return "" }
        return vm.localPrefix(key: "state", for: state)
    }

    /// Evaluate a Lua expression with local props injected.
    func evaluate(_ expression: String) throws -> LuaValue {
        guard let vm else { throw LuaError.initializationFailed }
        let prefix = "\(propsPrefix())\n\(statePrefix())"
        return try vm.execute(prefix + "return " + expression)
    }

    // MARK: - Typed Resolve Methods

    func string(_ value: Value<String>?) -> String {
        guard let value else { return "" }
        switch value {
        case .literal(let s): return s
        case .expression(let expr):
            guard vm != nil else { return "" }
            guard let result = try? evaluate(expr) else { return "" }
            switch result {
            case .string(let s): return s
            case .number(let n):
                if n == n.rounded() && n < 1e15 { return String(Int(n)) }
                return String(n)
            case .bool(let b): return b ? "true" : "false"
            case .nil: return ""
            default: return String(describing: result)
            }
        }
    }

    func number(_ value: Value<Double>?) -> Double? {
        guard let value else { return nil }
        switch value {
        case .literal(let n): return n
        case .expression(let expr):
            return (try? evaluate(expr))?.numberValue
        }
    }

    func integer(_ value: Value<Int>?) -> Int? {
        guard let value else { return nil }
        switch value {
        case .literal(let n): return n
        case .expression(let expr):
            return (try? evaluate(expr))?.numberValue.map(Int.init)
        }
    }

    func bool(_ value: Value<Bool>?, default defaultValue: Bool) -> Bool {
        guard let value, vm != nil else { return defaultValue }
        switch value {
        case .literal(let b): return b
        case .expression(let expr):
            do {
                let result = try evaluate(expr)
                if case .bool(let b) = result { return b }
                return defaultValue
            } catch {
                return defaultValue
            }
        }
    }

    func visible(_ value: Value<Bool>?) -> Bool {
        guard let value else { return true }
        switch value {
        case .literal(let b): return b
        case .expression(let expr):
            guard vm != nil else { return true }
            do {
                let result = try evaluate(expr)
                switch result {
                case .bool(let b): return b
                case .nil: return false
                default: return true
                }
            } catch {
                return true
            }
        }
    }

    func disabled(_ value: Value<Bool>?) -> Bool {
        return bool(value, default: false)
    }

    func items(_ expression: String?) -> [LuaValue] {
        guard let expression, vm != nil else { return [] }
        do {
            let trimmed = expression.trimmingCharacters(in: .whitespacesAndNewlines)
            let result: LuaValue
            if trimmed.contains("\n") {
                let prefix = propsPrefix()
                result = try vm!.execute(prefix + expression)
            } else {
                result = try evaluate(expression)
            }
            if case .array(let items) = result {
                return items
            }
            return []
        } catch {
            return []
        }
    }

    func options(_ source: OptionsSource?) -> [OptionDefinition] {
        guard let source else { return [] }
        switch source {
        case .static(let arr):
            return arr
        case .expression(let expr):
            guard vm != nil else { return [] }
            do {
                let result = try evaluate(expr)
                if case .array(let items) = result {
                    return items.compactMap { item in
                        guard let t = item.tableValue,
                              let label = t["label"]?.stringValue,
                              let value = t["value"]?.stringValue else { return nil }
                        return OptionDefinition(label: label, value: value)
                    }
                }
                return []
            } catch {
                return []
            }
        }
    }

    func direction(_ value: Value<DirectionAxis>?, default defaultValue: DirectionAxis = .vertical) -> DirectionAxis {
        guard let value else { return defaultValue }
        switch value {
        case .literal(let dir): return dir
        case .expression(let expr):
            let resolved = string(.expression(expr))
            return DirectionAxis(rawValue: resolved)
        }
    }

    func alignment(_ value: Value<ViewAlignment>?, default defaultValue: ViewAlignment = .leading) -> ViewAlignment {
        guard let value else { return defaultValue }
        switch value {
        case .literal(let a): return a
        case .expression(let expr):
            let resolved = string(.expression(expr))
            return ViewAlignment(rawValue: resolved)
        }
    }

    /// Resolves all `Value` expressions in a style to `.literal` values.
    func style(_ style: ComponentStyle?) -> ComponentStyle? {
        guard var style, vm != nil else { return style }

        // Resolve color strings
        if case .expression = style.color { style.color = resolveStringValue(style.color) }
        if case .expression = style.backgroundColor { style.backgroundColor = resolveStringValue(style.backgroundColor) }
        if case .expression = style.borderColor { style.borderColor = resolveStringValue(style.borderColor) }

        // Resolve alignment
        if case .expression = style.alignment {
            style.alignment = .literal(alignment(style.alignment))
        }

        // Resolve all numeric Value<Double> expressions
        style.fontSize = resolveDoubleValue(style.fontSize)
        style.padding = resolveDoubleValue(style.padding)
        style.paddingTop = resolveDoubleValue(style.paddingTop)
        style.paddingBottom = resolveDoubleValue(style.paddingBottom)
        style.paddingLeft = resolveDoubleValue(style.paddingLeft)
        style.paddingRight = resolveDoubleValue(style.paddingRight)
        style.paddingHorizontal = resolveDoubleValue(style.paddingHorizontal)
        style.paddingVertical = resolveDoubleValue(style.paddingVertical)
        style.margin = resolveDoubleValue(style.margin)
        style.marginTop = resolveDoubleValue(style.marginTop)
        style.marginBottom = resolveDoubleValue(style.marginBottom)
        style.marginLeft = resolveDoubleValue(style.marginLeft)
        style.marginRight = resolveDoubleValue(style.marginRight)
        style.marginHorizontal = resolveDoubleValue(style.marginHorizontal)
        style.marginVertical = resolveDoubleValue(style.marginVertical)
        style.borderRadius = resolveDoubleValue(style.borderRadius)
        style.borderWidth = resolveDoubleValue(style.borderWidth)
        style.width = resolveDoubleValue(style.width)
        style.height = resolveDoubleValue(style.height)
        style.minWidth = resolveDoubleValue(style.minWidth)
        style.minHeight = resolveDoubleValue(style.minHeight)
        style.maxWidth = resolveDoubleValue(style.maxWidth)
        style.maxHeight = resolveDoubleValue(style.maxHeight)
        style.spacing = resolveDoubleValue(style.spacing)
        style.opacity = resolveDoubleValue(style.opacity)
        style.cornerRadius = resolveDoubleValue(style.cornerRadius)
        style.scale = resolveDoubleValue(style.scale)
        style.rotation = resolveDoubleValue(style.rotation)
        style.aspectRatio = resolveDoubleValue(style.aspectRatio)
        style.layoutPriority = resolveDoubleValue(style.layoutPriority)

        // Resolve lineLimit
        if case .expression(let expr) = style.lineLimit {
            if let n = (try? evaluate(expr))?.numberValue {
                style.lineLimit = .literal(Int(n))
            }
        }

        return style
    }

    func transition(_ value: Value<String>?) -> AnyTransition {
        let str = string(value)
        guard !str.isEmpty else { return .opacity }
        let parts = str.lowercased().split(separator: ".")
        var result: AnyTransition? = nil
        var i = 0
        while i < parts.count {
            let t: AnyTransition
            switch parts[i] {
            case "opacity": t = .opacity
            case "slide": t = .slide
            case "scale": t = .scale
            case "move":
                i += 1
                let edge: Edge = i < parts.count ? (parts[i] == "top" ? .top : parts[i] == "leading" ? .leading : parts[i] == "trailing" ? .trailing : .bottom) : .bottom
                t = .move(edge: edge)
            default: t = .opacity
            }
            result = result.map { $0.combined(with: t) } ?? t
            i += 1
        }
        return result ?? .opacity
    }

    // MARK: - Private Helpers

    private func resolveDoubleValue(_ value: Value<Double>?) -> Value<Double>? {
        guard let value else { return nil }
        if case .expression(let expr) = value {
            if let n = (try? evaluate(expr))?.numberValue {
                return .literal(n)
            }
        }
        return value
    }

    private func resolveStringValue(_ value: Value<String>?) -> Value<String>? {
        guard let value else { return nil }
        if case .expression(let expr) = value {
            return .literal(string(.expression(expr)))
        }
        return value
    }
}

File: Android/melody-runtime/src/main/java/com/melody/runtime/renderer/ExpressionResolver.kt
package com.melody.runtime.renderer

import com.melody.core.schema.*
import com.melody.runtime.engine.LuaVM
import com.melody.runtime.engine.LuaValue

/**
 * Centralizes all Lua expression resolution.
 * Mirrors iOS `ExpressionResolver.swift`.
 */
class ExpressionResolver(
    private val vm: LuaVM?,
    private val props: Map<String, LuaValue>?
) {
    private fun propsPrefix(): String {
        return if (vm != null && props != null) vm.propsPrefix(props) else ""
    }

    private fun evaluate(expression: String): LuaValue {
        val prefix = propsPrefix()
        return vm!!.execute("${prefix}return $expression")
    }

    // MARK: - Typed Resolve Methods

    fun string(value: Value<String>?): String {
        if (value == null) return ""
        return when (value) {
            is Value.Literal -> value.value
            is Value.Expression -> {
                if (vm == null) return ""
                try {
                    when (val result = evaluate(value.expr)) {
                        is LuaValue.StringVal -> result.value
                        is LuaValue.NumberVal -> {
                            val n = result.value
                            if (n == n.toLong().toDouble() && n < 1e15) n.toLong().toString() else n.toString()
                        }
                        is LuaValue.BoolVal -> if (result.value) "true" else "false"
                        is LuaValue.Nil -> ""
                        else -> result.toString()
                    }
                } catch (_: Exception) { "" }
            }
        }
    }

    fun number(value: Value<Double>?): Double? {
        if (value == null) return null
        return when (value) {
            is Value.Literal -> value.value
            is Value.Expression -> {
                try { evaluate(value.expr).numberValue } catch (_: Exception) { null }
            }
        }
    }

    fun integer(value: Value<Int>?): Int? {
        if (value == null) return null
        return when (value) {
            is Value.Literal -> value.value
            is Value.Expression -> {
                try { evaluate(value.expr).numberValue?.toInt() } catch (_: Exception) { null }
            }
        }
    }

    fun bool(value: Value<Boolean>?, default: Boolean): Boolean {
        if (value == null || vm == null) return default
        return when (value) {
            is Value.Literal -> value.value
            is Value.Expression -> {
                try {
                    val result = evaluate(value.expr)
                    (result as? LuaValue.BoolVal)?.value ?: default
                } catch (_: Exception) { default }
            }
        }
    }

    fun visible(value: Value<Boolean>?): Boolean {
        if (value == null) return true
        return when (value) {
            is Value.Literal -> value.value
            is Value.Expression -> {
                if (vm == null) return true
                try {
                    when (val result = evaluate(value.expr)) {
                        is LuaValue.BoolVal -> result.value
                        is LuaValue.Nil -> false
                        else -> true
                    }
                } catch (_: Exception) { true }
            }
        }
    }

    fun disabled(value: Value<Boolean>?): Boolean {
        return bool(value, default = false)
    }

    fun direction(value: Value<DirectionAxis>?, default: DirectionAxis = DirectionAxis.Vertical): DirectionAxis {
        if (value == null) return default
        return when (value) {
            is Value.Literal -> value.value
            is Value.Expression -> {
                val resolved = string(Value.Expression(value.expr))
                DirectionAxis.from(resolved)
            }
        }
    }

    fun alignment(value: Value<ViewAlignment>?, default: ViewAlignment = ViewAlignment.Leading): ViewAlignment {
        if (value == null) return default
        return when (value) {
            is Value.Literal -> value.value
            is Value.Expression -> {
                val resolved = string(Value.Expression(value.expr))
                ViewAlignment.from(resolved)
            }
        }
    }

    /** Resolves all Value expressions in a style to Literal values. */
    fun style(style: ComponentStyle?): ComponentStyle? {
        if (style == null || vm == null) return style
        val resolved = style.copy()

        // Resolve color strings
        resolved.color = resolveStringValue(style.color)
        resolved.backgroundColor = resolveStringValue(style.backgroundColor)
        resolved.borderColor = resolveStringValue(style.borderColor)

        // Resolve alignment
        if (style.alignment is Value.Expression) {
            resolved.alignment = Value.Literal(alignment(style.alignment))
        }

        // Resolve all numeric Value<Double> expressions
        resolved.fontSize = resolveDoubleValue(style.fontSize)
        resolved.padding = resolveDoubleValue(style.padding)
        resolved.paddingTop = resolveDoubleValue(style.paddingTop)
        resolved.paddingBottom = resolveDoubleValue(style.paddingBottom)
        resolved.paddingLeft = resolveDoubleValue(style.paddingLeft)
        resolved.paddingRight = resolveDoubleValue(style.paddingRight)
        resolved.paddingHorizontal = resolveDoubleValue(style.paddingHorizontal)
        resolved.paddingVertical = resolveDoubleValue(style.paddingVertical)
        resolved.margin = resolveDoubleValue(style.margin)
        resolved.marginTop = resolveDoubleValue(style.marginTop)
        resolved.marginBottom = resolveDoubleValue(style.marginBottom)
        resolved.marginLeft = resolveDoubleValue(style.marginLeft)
        resolved.marginRight = resolveDoubleValue(style.marginRight)
        resolved.marginHorizontal = resolveDoubleValue(style.marginHorizontal)
        resolved.marginVertical = resolveDoubleValue(style.marginVertical)
        resolved.borderRadius = resolveDoubleValue(style.borderRadius)
        resolved.borderWidth = resolveDoubleValue(style.borderWidth)
        resolved.width = resolveDoubleValue(style.width)
        resolved.height = resolveDoubleValue(style.height)
        resolved.minWidth = resolveDoubleValue(style.minWidth)
        resolved.minHeight = resolveDoubleValue(style.minHeight)
        resolved.maxWidth = resolveDoubleValue(style.maxWidth)
        resolved.maxHeight = resolveDoubleValue(style.maxHeight)
        resolved.spacing = resolveDoubleValue(style.spacing)
        resolved.opacity = resolveDoubleValue(style.opacity)
        resolved.cornerRadius = resolveDoubleValue(style.cornerRadius)
        resolved.scale = resolveDoubleValue(style.scale)
        resolved.rotation = resolveDoubleValue(style.rotation)
        resolved.aspectRatio = resolveDoubleValue(style.aspectRatio)
        resolved.layoutPriority = resolveDoubleValue(style.layoutPriority)

        // Resolve lineLimit
        if (style.lineLimit is Value.Expression) {
            val expr = (style.lineLimit as Value.Expression).expr
            try {
                val n = evaluate(expr).numberValue
                if (n != null) resolved.lineLimit = Value.Literal(n.toInt())
            } catch (_: Exception) {}
        }

        return resolved
    }

    // MARK: - Private Helpers

    private fun resolveDoubleValue(value: Value<Double>?): Value<Double>? {
        if (value == null) return null
        if (value is Value.Expression) {
            try {
                val n = evaluate(value.expr).numberValue
                if (n != null) return Value.Literal(n)
            } catch (_: Exception) {}
        }
        return value
    }

    private fun resolveStringValue(value: Value<String>?): Value<String>? {
        if (value == null) return null
        if (value is Value.Expression) {
            return Value.Literal(string(value))
        }
        return value
    }
}
